### PR TITLE
Optimize Backward Propogation for BatchNorm Layer.

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hey everyone,
This PR aims to use faster backward propogation for BatchNorm Layer. The reason this was divided into two parts was simple i.e. The previous PR that implemented mini-batchnorm was a blocker for my tasks which is now removed and this is simply an optimization. Also, This gives me very high gradients(0.9 - 1) so this needs some debugging.
[Reference 1](https://kratzert.github.io/2016/02/12/understanding-the-gradient-flow-through-the-batch-normalization-layer.html) -> Previous BatchNorm
[Reference 2](http://cthorey.github.io./backpropagation/) -> Fast BatchNorm.

The values for both batchnorm match and hence both are correct.

### CODE REVIEW

**Comment by saksham189** [comment]
Looks like the tests are failing. Can you also benchmark and see the performance improvement we get from this as compared to the old implementation?

**Comment by kartikdutt18** [comment]
> Looks like the tests are failing. Can you also benchmark and see the performance improvement we get from this as compared to the old implementation?

Sure, I can run a benchmark. The gradient is very high so only gradient tests are failing. I'll look into it.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 1]
```diff
@@ -173,35 +173,30 @@ void BatchNorm<InputDataType, OutputDataType>::Backward(
...
-  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+  const arma::mat invVariance = 1.0 / (variance + eps);
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = input.n_rows / size;
 
   g.set_size(arma::size(input));
-  arma::cube gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(),
-      input.n_rows / size, size, input.n_cols, false, false);
-  arma::cube gTemp(const_cast<arma::Mat<eT>&>(g).memptr(),
-      input.n_rows / size, size, input.n_cols, false, false);
```
Comment (mrityunjay-tripathi) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: This will modify `gy` if `gyTemp` is changed somewhere. Probably if someone wants to access `previousLayer.Delta()`, it will not give the actual result, or if it is used in skip connections it may create a problem. Maybe I am not understanding correctly but can you please clarify this?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: gY is passed as const, If I try modifying it will, it will throw an error anyway (which is not done). Also this prevents creating an auxiliary array(copy) to prevent space allocation.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: Can you please try this small program?
```cpp
#include<armadillo>

using namespace arma;

void test(const arma::mat& x,
          arma::mat& y)
{
  size_t splitRatio = 2;
  arma::cube xTemp(const_cast<arma::mat&>(x).memptr(), x.n_rows / splitRatio, x.n_cols, splitRatio, false, false);
  xTemp.zeros();
  
  y.set_size(xTemp.n_rows * xTemp.n_cols, xTemp.n_slices);
  for (size_t i = 0; i < xTemp.n_slices; ++i)
  {
    y.col(i) = arma::vectorise(xTemp.slice(i));
  }
}

int main()
{
  arma::mat x = arma::randu(10, 4);
  arma::mat y;
  x.print("x initially:");
  test(x, y);
  x.print("x finally:");
}

```

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: Right, But I didn't change gYTemp anywhere. 
```

#include<armadillo>

using namespace arma;

void test(const arma::mat& x,
          arma::mat& y)
{
  size_t splitRatio = 2;
  arma::cube xTemp(const_cast<arma::mat&>(x).memptr(), x.n_rows / splitRatio, x.n_cols, splitRatio, false, false);

  y.set_size(xTemp.n_rows * xTemp.n_cols, xTemp.n_slices);
  for (size_t i = 0; i < xTemp.n_slices; ++i)
  {
    y.col(i) = arma::vectorise(xTemp.slice(i));
  }
}

int main()
{
  arma::mat x = arma::randu(10, 4);
  arma::mat y;
  x.print("x initially:");
  test(x, y);
  x.print("x finally:");
}
```
This shouldn't be an issue, so If I make sure not to change it should that be issue ?
Also take a look at Convolution layer, Transposed Conv layers and many more.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: Ohh...wait! I was reading `gTemp` as `gyTemp`. I'm really sorry 😂. Maybe someone else can suggest better what is good practice here. In my opinion, you can leave if it is "known" that it is not modified.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp:182]: No worries.

**Comment by zoq** [comment]
Let's keep this one open.
